helix-core/src/main/java/org/apache/helix/customizedstate/CustomizedStateProvider.java

```diff
@@ -0,0 +1,88 @@
+package org.apache.helix.customizedstate;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.helix.HelixDataAccessor;
+import org.apache.helix.HelixException;
+import org.apache.helix.HelixManager;
+import org.apache.helix.PropertyKey;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.model.CustomizedState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Class for Helix customers to operate on customized state
+ */
+public class CustomizedStateProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(CustomizedStateProvider.class);
+  private final HelixManager _helixManager;
+  String _instanceName;
+
+  public CustomizedStateProvider(HelixManager helixManager, String instanceName) {
+    _helixManager = helixManager;
+    _instanceName = instanceName;
+  }
+
+  /**
+   * Update the customized state based on the resource name and partition name
+   */
+  public synchronized void updateCustomizedState(String customizedStateName, String resourceName,
+      String partitionName, Map<String, String> customizedState) {
+    HelixDataAccessor accessor = _helixManager.getHelixDataAccessor();
+    PropertyKey.Builder keyBuilder = accessor.keyBuilder();
+    PropertyKey propertyKey =
+        keyBuilder.customizedState(_instanceName, customizedStateName, resourceName);
+    ZNRecord record = new ZNRecord(resourceName);
+    Map<String, Map<String, String>> mapFields = new HashMap<>();
+    mapFields.put(partitionName, customizedState);
+    record.setMapFields(mapFields);
```

Comment:
This will override other partition's data, right?  You need to read existing ZNRecord back, merge your changes and update to ZK.

Reply:
Also, if I have 5 fields in my customized state map, but here I will like just to update 1 of them, do I need to specified other 4 fields again in this map? Your implementation seems requiring so.

Reply:
Using update actually will not override other partition's data. I tested in the unit test. It will only override the same key's.
For the fields update, I was actually thinking about it. The implementation right now only support full update. I can add partial update. It's be useful if there're many fields that actually do not change.